test/e2e/suite/flake_attempt.go

```diff
@@ -0,0 +1,101 @@
+package suite
+
+import (
+	"context"
+	"fmt"
+	"slices"
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+	. "github.com/onsi/ginkgo/v2"
+	"github.com/onsi/gomega"
+	. "github.com/onsi/gomega"
+)
+
+type FlakeAttempt struct {
+	logger logr.Logger
+}
+type FlakeRun struct {
+	DeferCleanup    func(func(context.Context), ...interface{})
+	RetryableExpect func(actual interface{}, extra ...interface{}) Assertion
+}
+
+type IgnorablePanic struct{}
+
+func (IgnorablePanic) GinkgoRecoverShouldIgnoreThisPanic() {}
+
+func (f *FlakeAttempt) It(ctx context.Context, description string, flakeAttempts int, body func(ctx context.Context, flakeRun FlakeRun)) {
+	var lastErr error
+	for attempt := range flakeAttempts {
+		// register globally as well in case the test fails for any reason
+		// including being cancelled while this block is executing
+		// track if the cleanup runs and do not run it again if it does
+		cleanups := []func(context.Context){}
+		deferCleanup := func(cleanup func(context.Context), args ...interface{}) {
+			ran := false
+			onced := func(ctx context.Context) {
+				if ran {
+					f.logger.Info(fmt.Sprintf("Cleanup already ran for flake attempt %d, skipping", attempt+1))
+					return
+				}
+				ran = true
+				cleanup(ctx)
+			}
+			cleanups = append(cleanups, onced)
+			DeferCleanup(onced, args)
+		}
+
+		testGoMega := gomega.NewGomega(func(message string, callerSkip ...int) {
+			lastErr = fmt.Errorf("%s", message)
+			panic(IgnorablePanic{})
+		})
+		flakeRun := FlakeRun{
+			DeferCleanup:    deferCleanup,
+			RetryableExpect: testGoMega.Expect,
+		}
+
+		var panicableError error
+		var wg sync.WaitGroup
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			defer func() {
+				// if this was a ignorable panic, we ignore it
+				// to indicate that we should retry
+				// otherwise we store the panicable error to trigger
+				// the test to fail
+				e := recover()
+				if lastErr == nil && e != nil {
+					panicableError = e.(error)
```

Comment:
probably should check if this is an error? i know it should be, but if something else panic with something but an error, i think this will panic